test/opt/test_instruction.cpp

```diff
@@ -0,0 +1,118 @@
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "opt/instruction.h"
+
+#include "gmock/gmock.h"
+
+#include "spirv-tools/libspirv.h"
+#include "UnitSPIRV.h"
+
+namespace {
+
+using spvtest::MakeInstruction;
+using spvtools::ir::Instruction;
+using spvtools::ir::Operand;
+using ::testing::Eq;
+
+TEST(InstructionTest, CreateTrivial) {
+  Instruction empty;
+  EXPECT_EQ(SpvOpNop, empty.opcode());
+  EXPECT_EQ(0u, empty.type_id());
+  EXPECT_EQ(0u, empty.result_id());
+  EXPECT_EQ(0u, empty.NumOperands());
+  EXPECT_EQ(0u, empty.NumOperandWords());
+  EXPECT_EQ(0u, empty.NumInOperandWords());
+  EXPECT_EQ(empty.cend(), empty.cbegin());
+}
+
+TEST(InstructionTest, CreateWithOpcodeAndNoOperands) {
+  Instruction inst(SpvOpReturn);
+  EXPECT_EQ(SpvOpReturn, inst.opcode());
+  EXPECT_EQ(0u, inst.type_id());
+  EXPECT_EQ(0u, inst.result_id());
+  EXPECT_EQ(0u, inst.NumOperands());
+  EXPECT_EQ(0u, inst.NumOperandWords());
+  EXPECT_EQ(0u, inst.NumInOperandWords());
+  EXPECT_EQ(inst.cend(), inst.cbegin());
+}
+
+TEST(InstructionTest, CreateWithOpcodeAndOperands) {
+  auto i32 = MakeInstruction(SpvOpTypeInt, {44, 32, 1});
+  spv_parsed_operand_t parsed_operands[] = {
+      {1, 1, SPV_OPERAND_TYPE_RESULT_ID, SPV_NUMBER_NONE, 0},
+      {2, 1, SPV_OPERAND_TYPE_LITERAL_INTEGER, SPV_NUMBER_UNSIGNED_INT, 32},
+      {3, 1, SPV_OPERAND_TYPE_LITERAL_INTEGER, SPV_NUMBER_UNSIGNED_INT, 1},
+  };
+  spv_parsed_instruction_t parsed_inst{i32.data(),
+                                       uint16_t(i32.size()),
+                                       uint16_t(SpvOpTypeInt),
+                                       SPV_EXT_INST_TYPE_NONE,
+                                       0,   // type id
+                                       i32[1],  // result id
+                                       parsed_operands,
+                                       3};
+
+  Instruction inst(parsed_inst);
+  EXPECT_EQ(SpvOpTypeInt, inst.opcode());
+  EXPECT_EQ(0u, inst.type_id());
+  EXPECT_EQ(44u, inst.result_id());
+  EXPECT_EQ(3u, inst.NumOperands());
+  EXPECT_EQ(3u, inst.NumOperandWords());
+  EXPECT_EQ(2u, inst.NumInOperandWords());
+  auto cbegin = inst.cbegin();
+  auto cend = inst.cend();
+  EXPECT_NE(cend, inst.cbegin());
+
+  // Spot check iteration across operands.
+
+  auto iter = inst.cbegin();
+  const auto& operand0 = *iter;
```

Comment:
Some advice tells you not to have too much flow control in your tests.  I rewrote it your way anyway.